File: src/test/scala/parse/dont/validate/Generators.scala
package parse.dont.validate

import org.scalacheck.Arbitrary
import org.scalacheck.Gen.{choose, frequency, oneOf}
import parse.dont.validate.NIRDomain.*

object Generators {
  def validNIRGenerator(): Arbitrary[NIR] =
    Arbitrary {
      for {
        sex <- oneOf(Sex.values)
        year <- choose(1, 99)
        month <- oneOf(Month.values)
        department <- frequency(
          (9, choose(1, 95)),
          (1, 99)
        )
        city <- choose(1, 999)
        serialNumber <- choose(1, 999)
      } yield new NIR(
        sex,
        Year(year),
        month,
        Department(department),
        City(city),
        SerialNumber(serialNumber)
      )
    }
}

File: solution/kotlin/src/test/kotlin/parse/dont/validate/Generators.kt
package parse.dont.validate

import io.kotest.property.Arb
import io.kotest.property.arbitrary.*
import parse.dont.validate.NIR.*

object Generators {
    val nirGenerator = arbitrary { _ ->
        NIR(
            sex = Arb.enum<Sex>().bind(),
            year = Year(Arb.int(0, 99).bind()),
            month = Arb.enum<Month>().bind(),
            department = Department(
                Arb.frequency(
                    9 to Arb.int(1, 95),
                    1 to Arb.int(99, 99)
                ).bind()
            ),
            city = City(Arb.int(1, 999).bind()),
            serialNumber = SerialNumber(Arb.int(1, 999).bind())
        )
    }
}